source/extensions/filters/http/cors/cors_filter.cc

```diff
@@ -59,31 +74,32 @@ Http::FilterHeadersStatus CorsFilter::decodeHeaders(Http::RequestHeaderMap& head
     return Http::FilterHeadersStatus::Continue;
   }
 
-  const auto requestMethod = headers.AccessControlRequestMethod();
-  if (requestMethod == nullptr || requestMethod->value().empty()) {
+  const auto request_method = headers.getInline(access_control_request_method.handle());
+  if (request_method == nullptr || request_method->value().empty()) {
     return Http::FilterHeadersStatus::Continue;
   }
 
   auto response_headers{Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
       {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}})};
 
-  response_headers->setAccessControlAllowOrigin(origin_->value().getStringView());
+  response_headers->setInline(access_control_allow_origin.handle(),
+                              origin_->value().getStringView());
 
   if (allowCredentials()) {
-    response_headers->setReferenceAccessControlAllowCredentials(
-        Http::Headers::get().CORSValues.True);
+    response_headers->setReferenceInline(access_control_allow_credentials.handle(),
```

Comment:
This looks really pretty usable. How far does this take us towards being able to completely eliminate non-O(1) headers? 

Presumably there will always be some places we need to deal with dynamic headers, e.g. cluster header routing and the like. Are these going to be relatively few?

I still remain curious on whether the most basic flat hash map with ordering support in the value would be competitive. Maybe with an additional map for handles similar to what exists here to improve performance for the well-known header case.

Reply:
> Presumably there will always be some places we need to deal with dynamic headers, e.g. cluster header routing and the like. Are these going to be relatively few?

It really depends on the use case. For people that are dealing with a lot of dynamic config, especially that they don't control, it may not be reasonable to pre-define all of the "fast" headers they want. This would either necessitate a different implementation entirely, or an escape hatch where we build a map on-demand like @jmarantz prototyped.

> I still remain curious on whether the most basic flat hash map with ordering support in the value would be competitive. Maybe with an additional map for handles similar to what exists here to improve performance for the well-known header case.

I don't know. As we have discussed a bunch of times there are a lot of trade-offs here. The reason this change is appealing to me is a) it cleans up a bunch of stuff and b) it should be better in all cases than what we have now (pending benchmark verification). Later, we can get rid of it all, have pluggable implementations, add a dynamic hash creation setup, etc.

Reply:
I'm good with this PR as a strict improvement for the existing implementation for reasons (a) and (b). I think this thread could be used for @adisuissa to track for future directions.

Reply:
OK, @jmarantz @asraa @adisuissa any objections? If not I will start the cleanup and add some benchmarks, etc.

Reply:
This looks great from the filter's perspective -- no objections! I'm happy that some calls for `get()` in filters can now be replaced with something O(1).

A naive question though, basically all the compiled extensions will have their O(1) headers registered, even if at run-time it may not be used -- is that correct?

Reply:
> A naive question though, basically all the compiled extensions will have their O(1) headers registered, even if at run-time it may not be used -- is that correct?

Yes, basically every registered header will cost 8 bytes in the header map (only the specific map it's registered to) + the time to memset it to 0, whether it's used or not. It's identical to the current system but fully customizable. This means the default map when I'm done will exclude a bunch of headers only used by extensions.